### PR TITLE
Checked minimum version for lodestar

### PR DESCRIPTION
**Pull request by dsimog01**
For adding lodestar to the stakers UI, we need to check that each of the items in the stakers include the necessary tokens to communicate to lodestar (like JWT)

### CODE REVIEW

File: packages/dappmanager/src/modules/stakerConfig/setStakerConfig.ts [hunk 1]
```diff
@@ -508,3 +520,70 @@ function setStakerConfigOnDb<T extends Network>({
...
 }
+
+function checkLodestarMinVersions(
+  lodestarDnpName: string,
+  selectedExecClient: StakerItemOk<Network, "execution"> | undefined,
+  isSignerSelected: boolean | undefined,
+  web3signerPkg: InstalledPackageDataApiReturn | undefined
+): void {
+  if (!selectedExecClient) {
+    throw Error(`Execution client is not selected.`);
```
Comment (pablomendezroyo) [inline packages/dappmanager/src/modules/stakerConfig/setStakerConfig.ts:531]: This is check is wrong, the user may unselect the execution client which means that this value will be empty/undefined

File: packages/dappmanager/src/modules/stakerConfig/setStakerConfig.ts [hunk 2]
```diff
@@ -125,6 +127,16 @@ export async function setStakerConfig<T extends Network>({
       `Web3signer version ${currentWeb3signerPkg.version} is lower than the minimum version ${web3signer.minVersion} required to work with the stakers UI. Update it to continue.`
     );
 
+  // If consensus client is lodestar, check that all the execution client and signer versions
+  if (stakerConfig.consensusClient?.dnpName.includes("lodestar")) {
```
Comment (pablomendezroyo) [inline packages/dappmanager/src/modules/stakerConfig/setStakerConfig.ts:131]: I would add to this check that the execution client and signer exist

File: packages/dappmanager/src/params.ts [hunk 3]
```diff
@@ -30,6 +34,30 @@ if (process.env.TEST) {
 /** Absolute global ENVs .env file from DAPPMANAGER containers */
 const GLOBAL_ENVS_PATH = path.join(DNCORE_DIR, GLOBAL_ENVS_FILE_NAME);
 
+//Lodestar minimum versions
+const lodestarStakersMinimumVersions: {
+  [key: DnpName]: { [key: DnpName]: string };
```
Comment (pablomendezroyo) [inline packages/dappmanager/src/params.ts:39]: You can take advantage from the existing types for execution and consensus client types depending on the network:
```
ExecutionClient<Network>
ConsensusClient<Network>
```

File: packages/dappmanager/src/modules/stakerConfig/setStakerConfig.ts [hunk 4]
```diff
@@ -508,3 +520,70 @@ function setStakerConfigOnDb<T extends Network>({
...
+  if (!lodestarMinVersions) {
+    throw Error(
+      `Lodestar ${lodestarDnpName} minimum versions are not defined.`
+    );
+  }
+
+  const selectedExecClientVersion = selectedExecClient.data?.metadata.version;
+
+  //Throw error if execution client version cannot be fetched
+  if (selectedExecClientVersion === undefined) {
```
Comment (pablomendezroyo) [inline packages/dappmanager/src/modules/stakerConfig/setStakerConfig.ts:546]: Why not `if(!selectedExecClientVersion)` This way you check for falsy values also

File: packages/dappmanager/src/modules/stakerConfig/setStakerConfig.ts [hunk 5]
```diff
@@ -508,3 +520,70 @@ function setStakerConfigOnDb<T extends Network>({
...
+  }
+
+  checkStakersItemLodestarCompatibility({
+    version: selectedSignerVersion,
+    minVersion: lodestarMinVersions[web3signerPkg.dnpName],
+    dnpName: web3signerPkg.dnpName
+  });
+}
+
+function checkStakersItemLodestarCompatibility({
```
Comment (pablomendezroyo) [inline packages/dappmanager/src/modules/stakerConfig/setStakerConfig.ts:575]: I do not see the point of having a function that contains on it another function. You should use the `semver.lt` directly where required

File: packages/dappmanager/src/modules/stakerConfig/setStakerConfig.ts [hunk 6]
```diff
@@ -508,3 +520,70 @@ function setStakerConfigOnDb<T extends Network>({
...
+  const selectedExecClientVersion = selectedExecClient.data?.metadata.version;
+
+  //Throw error if execution client version cannot be fetched
+  if (selectedExecClientVersion === undefined) {
+    throw Error(
+      `Execution client ${selectedExecClient.dnpName} version could not be fetched.`
+    );
+  }
+
+  checkStakersItemLodestarCompatibility({
```
Comment (pablomendezroyo) [inline packages/dappmanager/src/modules/stakerConfig/setStakerConfig.ts:552]: Use `lt` directly

File: packages/dappmanager/src/modules/stakerConfig/setStakerConfig.ts [hunk 7]
```diff
@@ -508,3 +520,70 @@ function setStakerConfigOnDb<T extends Network>({
...
+
+  const selectedSignerVersion = web3signerPkg.version;
+
+  if (!selectedSignerVersion) {
+    throw Error(
+      `Signer ${web3signerPkg.dnpName} version could not be fetched.`
+    );
+  }
+
+  checkStakersItemLodestarCompatibility({
```
Comment (pablomendezroyo) [inline packages/dappmanager/src/modules/stakerConfig/setStakerConfig.ts:568]: Use `lt` directly

**Review by pablomendezroyo — CHANGES_REQUESTED** [review]

File: packages/dappmanager/src/modules/stakerConfig/setStakerConfig.ts (hunk 1 above)
Comment (dsimog01) [reply packages/dappmanager/src/modules/stakerConfig/setStakerConfig.ts:531]: But it should only enter the function if a CC is selected (lodestar) and this means that you also need a EC, right @pablomendezroyo ?

File: packages/dappmanager/src/modules/stakerConfig/setStakerConfig.ts (hunk 2 above)
Comment (dsimog01) [reply packages/dappmanager/src/modules/stakerConfig/setStakerConfig.ts:131]: Okay, I can add this

File: packages/dappmanager/src/modules/stakerConfig/setStakerConfig.ts (hunk 1 above)
Comment (pablomendezroyo) [reply packages/dappmanager/src/modules/stakerConfig/setStakerConfig.ts:531]: Then the function should be called only if both clients exists and you do not have to re-check it inside the function

**Review by pablomendezroyo — APPROVED** [review]

File: packages/dappmanager/src/modules/stakerConfig/params.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+  LodestarStakersMinimumVersions
+} from "./types";
+
+const lodestarStakersPraterMinimumVersions: ExecutionClientOrSignerVersion<"prater"> =
+  {
+    "goerli-nethermind.dnp.dappnode.eth": "1.0.8",
+    "goerli-geth.dnp.dappnode.eth": "0.4.27",
+    "goerli-erigon.dnp.dappnode.eth": "0.1.4",
+    "goerli-besu.dnp.dappnode.eth": "0.1.1",
+    "web3signer-prater.dnp.dappnode.eth": "0.1.17"
```
Comment (nflaig) [inline packages/dappmanager/src/modules/stakerConfig/params.ts:13]: `0.1.17` does not work with Lodestar as brain does send the correct API, need https://github.com/dappnode/StakingBrain/pull/196 to be merged, released and web3signer to be updated

File: packages/dappmanager/src/modules/stakerConfig/params.ts [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+const lodestarStakersGnosisMinimumVersions: ExecutionClientOrSignerVersion<"gnosis"> =
+  {
+    "nethermind-xdai.dnp.dappnode.eth": "1.0.25",
+    "web3signer-gnosis.dnp.dappnode.eth": "0.1.12"
+  };
+
+const lodestarStakersMainnetMinimumVersions: ExecutionClientOrSignerVersion<"mainnet"> =
+  {
+    "nethermind.public.dappnode.eth": "1.0.32",
+    "geth.dnp.dappnode.eth": "0.1.37",
```
Comment (nflaig) [inline packages/dappmanager/src/modules/stakerConfig/params.ts:25]: `0.1.37` does not support lodestar yet

```suggestion
    "geth.dnp.dappnode.eth": "0.1.38",
```

File: packages/dappmanager/src/modules/stakerConfig/params.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    "erigon.dnp.dappnode.eth": "0.1.37",
+    "besu.public.dappnode.eth": "1.2.7",
+    "web3signer.dnp.dappnode.eth": "0.1.6"
```
Comment (nflaig) [inline packages/dappmanager/src/modules/stakerConfig/params.ts:28]: this package does not support Lodestar yet, there are 2 open PRs https://github.com/dappnode/DAppNodePackage-web3signer/pulls?q=lodestar

**Review by nflaig — CHANGES_REQUESTED** [review]

File: packages/dappmanager/src/params.ts [hunk 11]
```diff
@@ -1,7 +1,9 @@
 import path from "path";
 import { FileFormat } from "./types.js";
 import { Architecture } from "@dappnode/dappnodesdk";
-import { EthClientTargetPackage, UserSettings } from "@dappnode/common";
+import {
+  EthClientTargetPackage,
+  UserSettings} from "@dappnode/common";
```
Comment (nflaig) [inline packages/dappmanager/src/params.ts:6]: formatting here looks a bit off

File: packages/dappmanager/src/modules/stakerConfig/params.ts (hunk 10 above)
Comment (dsimog01) [reply packages/dappmanager/src/modules/stakerConfig/params.ts:28]: Merged to web3signer

File: packages/dappmanager/src/modules/stakerConfig/params.ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    "nethermind-xdai.dnp.dappnode.eth": "1.0.25",
+    "web3signer-gnosis.dnp.dappnode.eth": "0.1.12"
+  };
+
+const lodestarStakersMainnetMinimumVersions: ExecutionClientOrSignerVersion<"mainnet"> =
+  {
+    "nethermind.public.dappnode.eth": "1.0.32",
+    "geth.dnp.dappnode.eth": "0.1.37",
+    "erigon.dnp.dappnode.eth": "0.1.37",
+    "besu.public.dappnode.eth": "1.2.7",
```
[the remark this answers is not in the document]
Comment (dsimog01) [reply packages/dappmanager/src/modules/stakerConfig/params.ts:27]: Merged!

File: packages/dappmanager/src/modules/stakerConfig/params.ts [hunk 13]
```diff
@@ -0,0 +1,36 @@
+import { Network } from "@dappnode/common";
+import {
+  ExecutionClientOrSignerVersion,
+  LodestarStakersMinimumVersions
+} from "./types";
+
+const lodestarStakersPraterMinimumVersions: ExecutionClientOrSignerVersion<"prater"> =
+  {
+    "goerli-nethermind.dnp.dappnode.eth": "1.0.8",
+    "goerli-geth.dnp.dappnode.eth": "0.4.27", // Needs publishment
```
Comment (nflaig) [inline packages/dappmanager/src/modules/stakerConfig/params.ts:10]: just noticed this geth version is wrong and acutally I think we do not even need to publish a new version as [v0.4.28](https://github.com/dappnode/DAppNodePackage-goerli-geth/releases/tag/v0.4.28) already supports Lodestar

```suggestion
    "goerli-geth.dnp.dappnode.eth": "0.4.28",
```

@dsimog01 on my dappnode I am running `0.4.27` and it tells me it is the latest version, was `0.4.28` not published correctly or I am missing something?

**Comment by nflaig** [comment]
@dsimog01 @pablomendezroyo Thank you guys for all the work! I think we are almost there to get Lodestar on the prater stackers UI. Is there anything I can assist on to push this to the finish line?

File: packages/dappmanager/src/modules/stakerConfig/params.ts [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+    "web3signer-gnosis.dnp.dappnode.eth": "0.1.13" // Needs publishment
+  };
+
+const lodestarStakersMainnetMinimumVersions: ExecutionClientOrSignerVersion<"mainnet"> =
+  {
+    "nethermind.public.dappnode.eth": "1.0.32",
+    "geth.dnp.dappnode.eth": "0.1.38", // Needs publishment
+    "erigon.dnp.dappnode.eth": "0.1.37",
+    "besu.public.dappnode.eth": "1.2.7", // Needs publishment
+    "web3signer.dnp.dappnode.eth": "0.1.6" // Needs publishment
```
Comment (nflaig) [inline packages/dappmanager/src/modules/stakerConfig/params.ts:28]: testing mainnet package at the moment, just noticed 0.1.6 does not yet support lodestar

```suggestion
    "web3signer.dnp.dappnode.eth": "0.1.7" // Needs publishment
```

File: packages/dappmanager/src/modules/stakerConfig/params.ts [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+    "goerli-geth.dnp.dappnode.eth": "0.4.28",
+    "goerli-erigon.dnp.dappnode.eth": "0.1.5",
+    "goerli-besu.dnp.dappnode.eth": "0.1.1",
+    "web3signer-prater.dnp.dappnode.eth": "0.1.18"
+  };
+
+const lodestarStakersGnosisMinimumVersions: ExecutionClientOrSignerVersion<"gnosis"> =
+  {
+    "nethermind-xdai.dnp.dappnode.eth": "1.0.29", // Needs publishment
+    "web3signer-gnosis.dnp.dappnode.eth": "0.1.13" // Needs publishment
```
Comment (nflaig) [inline packages/dappmanager/src/modules/stakerConfig/params.ts:19]: 0.1.13 is already published but does not yet include the Lodestar api token

```suggestion
    "web3signer-gnosis.dnp.dappnode.eth": "0.1.14" // Needs publishment
```

**Review by pablomendezroyo — APPROVED** [review]

**Pull request merged**